polygon/bridge/bridge_test.go

```diff
@@ -0,0 +1,252 @@
+	require.NoError(t, err)
+
+	require.Equal(t, 2, len(res))                             // have first two events
+	require.Equal(t, event1Data, rlp.RawValue(res[0].Data())) // check data fields
+	require.Equal(t, event2Data, rlp.RawValue(res[1].Data()))
+
+	res, err = b.GetEvents(ctx, 4)
+	require.NoError(t, err)
+
+	event3Data, err := event3.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	require.Equal(t, 1, len(res))
+	require.Equal(t, event3Data, rlp.RawValue(res[0].Data()))
+
+	// get non-sprint block
+	res, err = b.GetEvents(ctx, 1)
+	require.Error(t, err)
+
+	res, err = b.GetEvents(ctx, 3)
+	require.Error(t, err)
+
+	// check block 0
+	res, err = b.GetEvents(ctx, 0)
+	require.Error(t, err)
+
+	wg.Wait()
+}
+
+func TestBridge_Unwind(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	stateReceiverABI := bor.GenesisContractStateReceiverABI()
+	heimdallClient, b := setup(t, stateReceiverABI)
+
+	event1 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      1,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x01"),
+		},
+		Time: time.Unix(50, 0), // block 2
+	}
+	event2 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      2,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x02"),
+		},
+		Time: time.Unix(100, 0), // block 2
+	}
+	event3 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      3,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x03"),
+		},
+		Time: time.Unix(200, 0), // block 4
+	}
+	event4 := &heimdall.EventRecordWithTime{
+		EventRecord: heimdall.EventRecord{
+			ID:      4,
+			ChainID: "80001",
+			Data:    hexutil.MustDecode("0x03"),
+		},
+		Time: time.Unix(300, 0), // block 6
+	}
+
+	events := []*heimdall.EventRecordWithTime{event1, event2, event3, event4}
+
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(events, nil).Times(1)
+	heimdallClient.EXPECT().FetchStateSyncEvents(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return([]*heimdall.EventRecordWithTime{}, nil).AnyTimes()
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	go func(bridge bridge.Service) {
+		defer wg.Done()
+
+		err := bridge.Run(ctx)
+		if err != nil {
+			if !errors.Is(err, ctx.Err()) {
+				t.Error(err)
+			}
+
+			return
+		}
+	}(b)
+
+	err := b.Synchronize(ctx, &types.Header{Number: big.NewInt(100)}) // hack to wait for b.ready
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	blocks := getBlocks(t, 8)
+
+	err = b.ProcessNewBlocks(ctx, blocks)
+	require.NoError(t, err)
+
+	event3Data, err := event3.Pack(stateReceiverABI)
+	require.NoError(t, err)
+
+	res, err := b.GetEvents(ctx, 4)
+	require.Equal(t, event3Data, rlp.RawValue(res[0].Data()))
+	require.NoError(t, err)
+
+	err = b.Unwind(ctx, &types.Header{Number: big.NewInt(3)})
+
+	_, err = b.GetEvents(ctx, 4)
+	require.Error(t, err)
+
+	wg.Wait()
```

Comment:
think you may need to keep the `cancel` before wg.Wait? otherwise it wait won't terminate?